### PR TITLE
Enable YAML 1.2 support for non-ansible files

### PR DESCRIPTION
**Pull request by ssbarnea**
Fixes: #3790

### CODE REVIEW

File: test/test_yaml_utils.py [hunk 1]
```diff
@@ -202,8 +203,7 @@ def test_custom_ruamel_yaml_emitter(
     assert output == expected_output
 
 
-@pytest.fixture(name="yaml_formatting_fixtures")
-def fixture_yaml_formatting_fixtures(fixture_filename: str) -> tuple[str, str, str]:
```
Comment (cognifloyd) [inline test/test_yaml_utils.py:206]: I made this a fixture so that any issues here would show up as a test error instead of a test failure. Why did you change it to a helper function?

File: src/ansiblelint/yaml_utils.py [hunk 2]
```diff
@@ -817,8 +819,12 @@ def __init__(
                 - name: Task
         """
         # Default to reading/dumping YAML 1.1 (ruamel.yaml defaults to 1.2)
```
Comment (cognifloyd) [inline src/ansiblelint/yaml_utils.py:821]: This comment is out of date now, because this uses ruamel.yaml's default of 1.2. It should probably be moved to `transformer.py`.

File: src/ansiblelint/yaml_utils.py [hunk 3]
```diff
@@ -939,7 +950,11 @@ def version(self, value: str | tuple[int, int] | None) -> None:
         So, if a file does not include the directive, it sets this to None.
         But, None effectively resets the parsing version to YAML 1.2 (ruamel's default).
         """
-        self._yaml_version = value if value is not None else self._yaml_version_default
+        if value is not None:
+            self._yaml_version = value
+        elif hasattr(self, "_yaml_version_default"):
+            self._yaml_version = self._yaml_version_default
+        # We do nothing if the object did not had a previous default version defined
```
Comment (cognifloyd) [inline src/ansiblelint/yaml_utils.py:957]:
```suggestion
        # We do nothing if the object did not have a previous default version defined
```

File: src/ansiblelint/yaml_utils.py [hunk 4]
```diff
@@ -874,6 +880,9 @@ def __init__(
         # (see https://stackoverflow.com/a/44314840/1134951)
         # self.Representer.add_representer(
 
+        if version:
+            self.version = version
```
Comment (cognifloyd) [inline src/ansiblelint/yaml_utils.py:884]: Why did you add this? Setting `self.version` actually sets `self._yaml_version` which is already configured above. Why do we need to set it again?

File: src/ansiblelint/yaml_utils.py [hunk 5]
```diff
@@ -977,7 +992,11 @@ def dumps(self, data: Any) -> str:
                 stream.write(preamble_comment)
             self.dump(data, stream)
             text = stream.getvalue()
-        return self._post_process_yaml(text)
+        strip_version_directive = hasattr(self, "_yaml_version_default")
```
Comment (cognifloyd) [inline src/ansiblelint/yaml_utils.py:995]: Is this ever false? I thought ruamel.yaml had a default for `self._yaml_version_default`.

File: src/ansiblelint/transformer.py [hunk 6]
```diff
@@ -92,7 +96,9 @@ def run(self) -> None:
                 # We need a fresh YAML() instance for each load because ruamel.yaml
                 # stores intermediate state during load which could affect loading
                 # any other files. (Based on suggestion from ruamel.yaml author)
-                yaml = FormattedYAML()
+                yaml = FormattedYAML(
+                    version=(1, 1) if file.is_owned_by_ansible() else None,
```
Comment (cognifloyd) [inline src/ansiblelint/transformer.py:100]: The comment about the default version being 1.1 should probably move here.

**Review by cognifloyd — DISMISSED** [review]

File: src/ansiblelint/yaml_utils.py (hunk 5 above)
Comment (ssbarnea) [reply src/ansiblelint/yaml_utils.py:995]: That is there for a good reason, that attribute is not always defined!

File: test/test_yaml_utils.py (hunk 1 above)
Comment (ssbarnea) [reply test/test_yaml_utils.py:206]: It made the testing harder and also maintenance, vscode got bit confused by it. I also think that having a test failure is fine.

**Review by cidrblock — APPROVED** [review]

File: src/ansiblelint/transformer.py [hunk 7]
```diff
@@ -92,7 +96,12 @@ def run(self) -> None:
                 # We need a fresh YAML() instance for each load because ruamel.yaml
                 # stores intermediate state during load which could affect loading
                 # any other files. (Based on suggestion from ruamel.yaml author)
-                yaml = FormattedYAML()
+                yaml = FormattedYAML(
+                    # Ansible only uses YAML 1.1, but others files should use newer 1.2 (ruamel.yaml defaults to 1.2)
+                    version=(1, 1)
```
[the remark this answers is not in the document]
Comment (ssbarnea) [reply src/ansiblelint/transformer.py:101]: There reason why I wanted to fallback to null was because that would make it make it future-proof for when ruamel.yaml will switch to versions such 1.3 or more.

What we want is to override the default for ansible file, and not touch it for others.

File: src/ansiblelint/yaml_utils.py (hunk 5 above)
Comment (cognifloyd) [reply src/ansiblelint/yaml_utils.py:990]: Doh! Of course. I added `self._yaml_version` and `self._yaml_version_default` to work around ruamel.yaml's quirky use of the `version` attribute. This PR changes that so those are only present sometimes. Hmm...

File: src/ansiblelint/yaml_utils.py [hunk 8]
```diff
@@ -816,10 +818,12 @@ def __init__(
...
-        # Default to reading/dumping YAML 1.1 (ruamel.yaml defaults to 1.2)
-        self._yaml_version_default: tuple[int, int] = (1, 1)
-        self._yaml_version: str | tuple[int, int] = self._yaml_version_default
-
+        if version:
+            if isinstance(version, str):
+                x, y = version.split(".", maxsplit=1)
+                version = (int(x), int(y))
+            self._yaml_version_default: VersionType = version
+            self._yaml_version: VersionType = self._yaml_version_default
```
Comment (cognifloyd) [inline src/ansiblelint/yaml_utils.py:826]:
```suggestion
        if isinstance(version, str):
            x, y = version.split(".", maxsplit=1)
            version = (int(x), int(y))
        self._yaml_version_default: VersionType | None = version
        self._yaml_version: VersionType | None = self._yaml_version_default
        self._strip_yaml_version_directive = version is not None
```

File: src/ansiblelint/yaml_utils.py [hunk 9]
```diff
@@ -920,16 +924,18 @@ def _defaults_from_yamllint_config() -> dict[str, bool | int | str]:
...
 
         Ansible uses PyYAML which only supports YAML 1.1. ruamel.yaml defaults to 1.2.
         So, we have to make sure we dump yaml files using YAML 1.1.
         We can relax the version requirement once ansible uses a version of PyYAML
         that includes this PR: https://github.com/yaml/pyyaml/pull/555
         """
-        return self._yaml_version
+        if hasattr(self, "_yaml_version"):
+            return self._yaml_version
+        return None
```
Comment (cognifloyd) [inline src/ansiblelint/yaml_utils.py:938]:
```suggestion
        return self._yaml_version
```

File: src/ansiblelint/yaml_utils.py [hunk 10]
```diff
@@ -939,7 +945,11 @@ def version(self, value: str | tuple[int, int] | None) -> None:
         So, if a file does not include the directive, it sets this to None.
         But, None effectively resets the parsing version to YAML 1.2 (ruamel's default).
         """
-        self._yaml_version = value if value is not None else self._yaml_version_default
+        if value is not None:
+            self._yaml_version = value
+        elif hasattr(self, "_yaml_version_default"):
+            self._yaml_version = self._yaml_version_default
+        # We do nothing if the object did not have a previous default version defined
```
Comment (cognifloyd) [inline src/ansiblelint/yaml_utils.py:952]: I like the ternary expression that was here before. But this would work too:

```suggestion
        if value is not None:
            self._yaml_version = value
        else:
            self._yaml_version = self._yaml_version_default
```

File: src/ansiblelint/yaml_utils.py [hunk 11]
```diff
@@ -977,7 +987,11 @@ def dumps(self, data: Any) -> str:
                 stream.write(preamble_comment)
             self.dump(data, stream)
             text = stream.getvalue()
-        return self._post_process_yaml(text)
+        strip_version_directive = hasattr(self, "_yaml_version_default")
+        return self._post_process_yaml(
+            text,
+            strip_version_directive=strip_version_directive,
+        )
```
Comment (cognifloyd) [inline src/ansiblelint/yaml_utils.py:994]:
```suggestion
        return self._post_process_yaml(
            text,
            strip_version_directive=self._strip_yaml_version_directive,
        )
```

**Review by cognifloyd — CHANGES_REQUESTED** [review]
The purpose of relying on the presence/absence of the `_yaml_version*` attributes was not clear to me. What about using a separate `_strip_yaml_version_directive` attribute to define that explicitly instead of implying it using `hasattr`?

File: src/ansiblelint/transformer.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (cognifloyd) [reply src/ansiblelint/transformer.py:101]: Agreed. We should stick with `None` as the default value.

File: src/ansiblelint/yaml_utils.py (hunk 10 above)
Comment (cognifloyd) [reply src/ansiblelint/yaml_utils.py:952]: So, this would be my preference for this block:
```suggestion
        self._yaml_version = value if value is not None else self._yaml_version_default
```

File: src/ansiblelint/yaml_utils.py (hunk 8 above)
Comment (ssbarnea) [reply src/ansiblelint/yaml_utils.py:826]: I do not want to touch this right now, especially as I know that a newer patch of the library introduces some data that we might want to use, look at 0.18.4 changelog entries https://pypi.org/project/ruamel.yaml/

We can do extra polishing in follow-up, but I prefer to use whatever they have instead of setting new properties that have meaning only for our subclass. I need to play with it bit more too see what data I have available.
